src/LDATaskS3.cpp

```diff
@@ -0,0 +1,385 @@
+#include <Tasks.h>
+
+#include "PSSparseServerInterface.h"
+#include "S3Client.h"
+#include "S3SparseIterator.h"
+#include "Serializers.h"
+#include "Utils.h"
+
+#include <pthread.h>
+#include <thread>
+
+#undef DEBUG
+
+namespace cirrus {
+
+void LDATaskS3::push_gradient(char* gradient_mem,
+                              int total_sampled_tokens,
+                              int total_sampled_docs,
+                              uint32_t to_send_size) {
+#ifdef DEBUG
+  auto before_push_us = get_time_us();
+  std::cout << "Publishing gradients" << std::endl;
+#endif
+  psint->send_lda_update(gradient_mem, total_sampled_tokens, total_sampled_docs,
+                         to_send_size + sizeof(int) * 2);
+#ifdef DEBUG
+  std::cout << "Published gradients!" << std::endl;
+  auto elapsed_push_us = get_time_us() - before_push_us;
+  static uint64_t before = 0;
+  if (before == 0)
+    before = get_time_us();
+  auto now = get_time_us();
+  std::cout << "[WORKER] "
+            << "Worker task published gradient"
+            << " with version: " << gradient->getVersion()
+            << " at time (us): " << get_time_us()
+            << " took(us): " << elapsed_push_us << " bw(MB/s): " << std::fixed
+            << (1.0 * gradient->getSerializedSize() / elapsed_push_us / 1024 /
+                1024 * 1000 * 1000)
+            << " since last(us): " << (now - before) << "\n";
+  before = now;
+#endif
+}
+
+void LDATaskS3::upload_wih_bucket_id_fn(char* mem_to_send,
+                                        uint64_t to_send_size,
+                                        int& upload_lock,
+                                        int bucket_id) {
+  while (true) {
```

Comment:
Don't understand this code. As far as I can understand this task is single threaded (help_upload_threads is not used).

Reply:
^^